presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashBasedPartialDistinctLimit.java

```diff
@@ -0,0 +1,102 @@
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.DistinctLimitNode;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static java.util.Objects.requireNonNull;
+
+public class HashBasedPartialDistinctLimit
+        implements PlanOptimizer
+{
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public HashBasedPartialDistinctLimit(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(variableAllocator, "variableAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+        if (SystemSessionProperties.isHashBasedDistinctLimitEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, idAllocator, variableAllocator, functionAndTypeManager), plan, null);
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final PlanNodeIdAllocator idAllocator;
+        private final PlanVariableAllocator variableAllocator;
+        private final FunctionAndTypeManager functionAndTypeManager;
+
+        private Rewriter(Session session, PlanNodeIdAllocator idAllocator, PlanVariableAllocator variableAllocator, FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.session = session;
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+            this.functionAndTypeManager = functionAndTypeManager;
+        }
+
+        @Override
+        public PlanNode visitDistinctLimit(DistinctLimitNode node, RewriteContext<Void> context)
+        {
+            if (node.isPartial() && node.getLimit() <= SystemSessionProperties.getHashBasedDistinctLimitThreshold(session)) {
```

Comment:
Using com.facebook.presto.sql.planner.iterative.Rule would make it a lot simpler and nicer.

Reply:
Harder to understand. I like visitor pattern better than regex-like rules.

Reply:
nits: static import

Reply:
Done